### PR TITLE
Add support for Hibernate5

### PR DESCRIPTION
**Pull request by shadanan**
Tests for Hibernate5 and Hibernate3, and examples, forthcoming in subsequent PRs

### CODE REVIEW

File: elide-dbmanager/elide-dbmanager-hibernate5/pom.xml [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    <artifactId>elide-dbmanager-hibernate5</artifactId>
+    <packaging>jar</packaging>
+    <name>Elide Hibernate5 Database Manager</name>
+    <description>Elide database manager for Hibernate5 support</description>
+    <parent>
+        <groupId>com.yahoo.elide</groupId>
+        <artifactId>elide-dbmanager-parent-pom</artifactId>
+        <version>1.0.0.1-SNAPSHOT</version>
+    </parent>
+
```
Comment (DeathByTape) [inline elide-dbmanager/elide-dbmanager-hibernate5/pom.xml]: Will need to add some fields here for maven central compliance. See: https://github.com/yahoo/elide/pull/13

You can probably just copy all of the blocks directly. Namely, you need the following:

- [ ] JavaDoc and Sources (just include plugins in section)
- [X] Signed artifacts (this is done for us by bintray)
- [ ] Add top-level URL
- [ ] Add license block
- [ ] Add developer block
- [ ] Add scm block

File: elide-dbmanager/elide-dbmanager-hibernate5/src/main/java/com/yahoo/elide/dbmanagers/hibernate5/HibernateManager.java [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+        private boolean inUse = false;
+        private boolean hasNext;
+
+        public ScrollableIterator(ScrollableResults scroll) {
+            this.scroll = scroll;
+        }
+
+        @Override
+        public Iterator<T> iterator() {
+            if (inUse) {
```
Comment (DeathByTape) [inline elide-dbmanager/elide-dbmanager-hibernate5/src/main/java/com/yahoo/elide/dbmanagers/hibernate5/HibernateManager.java]: This section concerns me a little bit (probably the same as hibernate3)? What happens if we have two threads `A` and `B` with a simultaneous access to iterator. That is, consider if threads A and B concurrently execute lines 64 - 68.

In such a scenario we wouldn't detect the concurrent modification. It feels like the `inUse` variable is a critical section to me.

File: elide-dbmanager/elide-dbmanager-hibernate5/src/test/java/com/yahoo/elide/dbmanagers/hibernate5/HibernateManagerTest.java [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+import static org.testng.Assert.*;
+
+public class HibernateManagerTest {
+    // TODO: Create tests for our manager
+    @Test
+    public void todo() {
+        System.err.println("Still need to write tests for hibernate3 manager!");
+    }
+}
```
Comment (DeathByTape) [inline elide-dbmanager/elide-dbmanager-hibernate5/src/test/java/com/yahoo/elide/dbmanagers/hibernate5/HibernateManagerTest.java]: new line. boo intellij always misses this with tests :-1:

File: elide-dbmanager/elide-dbmanager-hibernate5/src/main/java/com/yahoo/elide/dbmanagers/hibernate5/HibernateManager.java (hunk 2 above)
Comment (shadanan) [inline elide-dbmanager/elide-dbmanager-hibernate5/src/main/java/com/yahoo/elide/dbmanagers/hibernate5/HibernateManager.java]: Yeah. This file is copied and pasted from Hibernate3. I just modified it to work with Hibernate5.

Comment (DeathByTape) [inline elide-dbmanager/elide-dbmanager-hibernate5/src/main/java/com/yahoo/elide/dbmanagers/hibernate5/HibernateManager.java]: Just a follow-up: after talking in person, I think Shad has a good point here. Perhaps-- for performance-- we can simply throw the exception iff we detect the issue (i.e. this shouldn't be used concurrently), but we don't make any strong guarantees about detection.

File: elide-dbmanager/elide-dbmanager-hibernate5/src/main/java/com/yahoo/elide/dbmanagers/hibernate5/PersistenceManager.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+import java.io.IOException;
+
+/**
+ * Manager for javax.persistence compatible db resource
+ */
+public class PersistenceManager extends DatabaseManager {
+    private final EntityManagerFactory entityManagerFactory;
+
+    public PersistenceManager(EntityManagerFactory entityManagerFactory) {
+        this.entityManagerFactory = entityManagerFactory;
```
Comment (DeathByTape) [inline elide-dbmanager/elide-dbmanager-hibernate5/src/main/java/com/yahoo/elide/dbmanagers/hibernate5/PersistenceManager.java]: Should check entityManagerFactory. Maybe: `Preconditions.checkNotNull(entityManagerFactory);`

File: elide-dbmanager/elide-dbmanager-hibernate5/src/main/java/com/yahoo/elide/dbmanagers/hibernate5/PersistenceManager.java [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    }
+
+    @Override
+    public void populateEntityDictionary(EntityDictionary dictionary) {
+        for (EntityType entity : entityManagerFactory.getMetamodel().getEntities()) {
```
Comment (DeathByTape) [inline elide-dbmanager/elide-dbmanager-hibernate5/src/main/java/com/yahoo/elide/dbmanagers/hibernate5/PersistenceManager.java]: What if `getMetaModel()` or `getEntities()` returns `null`? Should we throw something more descriptive here than `NPE`?

Comment (shadanan) [inline elide-dbmanager/elide-dbmanager-hibernate5/src/main/java/com/yahoo/elide/dbmanagers/hibernate5/PersistenceManager.java]: It doesn't seem to be easy to get those to be null. I tried removing the classes from the persistence.xml file, and we just get back an empty entities set.

**Comment by DeathByTape** [comment]
:+1:

**Comment by yahoocla** [comment]
CLA is valid!

**Pull request merged**